src/uc_paymentwall.module

```diff
@@ -2,308 +2,324 @@
-
-        drupal_write_record('paymentwall_order_storage', $data);
-        unset($_SESSION['order']);
-        return $widget->getHtmlCode();
-      } else {
-        drupal_goto();
-        die();
-      }
-
-      break;
-  }
-}
+    if ($pingback->validate()) {
 
-/**
- * Ping back example.com/paymentwall_pingback
- */
-function uc_paymentwall_pingbackpayment() {
-  require 'paymentwall_api/lib/paymentwall.php';
-  $data = db_select('paymentwall_settings', 's')
-          ->fields('s', array('app_key', 'secret_key'))
-          ->execute()
-          ->fetch();
-
-  init_paymentwall_config($data);
-
-  $params = $_GET;
-  if (isset($params['q']) && $params['q'] != '') {
-      unset($params['q']);
-  }
-
-  $pingback = new Paymentwall_Pingback($params, $_SERVER['REMOTE_ADDR']);
-  if ($pingback->validate()) {
-    $result = db_select('paymentwall_order_storage', 'os')
-                ->fields('os')
-                ->condition('uid', $pingback->getUserId())
-                ->condition('oid',  $pingback->getProduct()->getId())
-                ->execute()
-                ->fetchAssoc();
-
-    $order = unserialize(base64_decode($result['order_obj']));
-
-    if($order) {
-      if ($pingback->isDeliverable()) {
-
-        $id = $pingback->getProduct()->getId();
-        $method = 'Paymentwall';
-        $amount = number_format($order->order_total, 2, '.', '');
-
-        if(is_int($pingback->getUserId())) {
-          $uid = $pingback->getUserId();
-        } else {
-          $uid = 0; // for anonymoys users
-        }
+        $order = uc_order_load($pingback->getProductId());
+
+        if ($order) {
+            if ($pingback->isDeliverable()) {
+
+                $method = 'Paymentwall';
+                $amount = number_format($order->order_total, 2, '.', '');
+
+                if (is_int($pingback->getUserId())) {
+                    $uid = $pingback->getUserId();
+                } else {
+                    $uid = 0; // for guest
+                }
+
+                if (variable_get('uc_pw_delivery')) {
+                    $delivery = new Paymentwall_GenerericApiObject('delivery');
+                    $response = $delivery->post(getDeliveryConfirmationData($order, $pingback->getReferenceId()));
+                }
 
-        $comment = 'User ' . $uid . ' complate purchase';
+                $comment = 'User ' . $uid . ' completed purchase';
 
-        uc_payment_enter($id, $method, $amount, $uid, null, $comment);
-        uc_cart_complete_sale($order, variable_get('uc_new_customer_login', FALSE));
+                uc_payment_enter($pingback->getProductId(), $method, $amount, $uid, NULL, $comment);
+                uc_cart_complete_sale($order, variable_get('uc_new_customer_login', FALSE));
 
-      } else if ($pingback->isCancelable()) {
-        uc_order_delete($pingback->getProduct()->getId());
-      }
+            } elseif ($pingback->isCancelable()) {
+                uc_order_delete($pingback->getProductId());
```

Comment:
Is that delete an order from the system?

If yes - it's wrong way, we should somehow mark this order as canceled